// frog/_lib/utils/requestToContext.js
import {} from 'hono';
import {} from '../types.js';
import { deserializeJson } from './deserializeJson.js';
import { fromQuery } from './fromQuery.js';
import * as jws from './jws.js';
import { verifyFrame } from './verifyFrame.js';
export async function requestToContext(req, { hubApiUrl, secret, verify = true }) {
    const { trustedData, untrustedData } = (await req.json().catch(() => { })) || {};
    const { initialPath, previousState, previousButtonValues } = await (async () => {
        if (untrustedData?.state) {
            const state = deserializeJson(untrustedData.state);
            if (secret && state.previousState)
                state.previousState = JSON.parse(await jws.verify(state.previousState, secret));
            return state;
        }
        if (req.query())
            return fromQuery(req.query());
        return {};
    })();
    const trustedFrameData = await (async () => {
        if (verify === false)
            return null;
        if (!trustedData)
            return null;
        if (!hubApiUrl)
            return null;
        try {
            const { frameData } = await verifyFrame({
                hubApiUrl,
                frameUrl: untrustedData.url,
                trustedData,
                url: req.url,
            });
            return { ...frameData, state: frameData.state || untrustedData.state };
        }
        catch (err) {
            if (verify === 'silent')
                return null;
            throw err;
        }
    })();
    return {
        initialPath: initialPath ? initialPath : new URL(req.url).pathname,
        previousState,
        previousButtonValues,
        frameData: trustedFrameData || untrustedData,
        status: req.method === 'POST' ? 'response' : 'initial',
        url: req.url,
        verified: Boolean(trustedFrameData),
    };
}
