// frog/_lib/components/Button.js
import { jsx as _jsx } from "frog/jsx/jsx-runtime";
ButtonRoot.__type = 'button';
export function ButtonRoot({ action, children, 
// @ts-ignore - private
index = 1, value, }) {
    return [
        _jsx("meta", { property: `fc:frame:button:${index}`, content: normalizeChildren(children), ...(value ? { 'data-value': value } : {}) }),
        _jsx("meta", { property: `fc:frame:button:${index}:action`, content: "post" }),
        action && (_jsx("meta", { property: `fc:frame:button:${index}:target`, content: action })),
    ];
}
ButtonLink.__type = 'button';
export function ButtonLink({ children, 
// @ts-ignore - private
index = 1, href, }) {
    return [
        _jsx("meta", { property: `fc:frame:button:${index}`, content: normalizeChildren(children), "data-value": "_l" }),
        _jsx("meta", { property: `fc:frame:button:${index}:action`, content: "link" }),
        _jsx("meta", { property: `fc:frame:button:${index}:target`, content: href }),
    ];
}
ButtonMint.__type = 'button';
export function ButtonMint({ children, 
// @ts-ignore - private
index = 1, target, }) {
    return [
        _jsx("meta", { property: `fc:frame:button:${index}`, content: normalizeChildren(children), "data-value": "_m" }),
        _jsx("meta", { property: `fc:frame:button:${index}:action`, content: "mint" }),
        _jsx("meta", { property: `fc:frame:button:${index}:target`, content: target }),
    ];
}
ButtonRedirect.__type = 'button';
export function ButtonRedirect({ children, 
// @ts-ignore - private
index = 1, location, }) {
    return [
        _jsx("meta", { property: `fc:frame:button:${index}`, content: normalizeChildren(children), "data-type": "redirect", "data-value": `_r:${location}` }),
        _jsx("meta", { property: `fc:frame:button:${index}:action`, content: "post_redirect" }),
        location && (_jsx("meta", { property: `fc:frame:button:${index}:target`, content: location })),
    ];
}
ButtonReset.__type = 'button';
export function ButtonReset({ children, 
// @ts-ignore - private
index = 1, }) {
    return (_jsx("meta", { property: `fc:frame:button:${index}`, content: normalizeChildren(children), "data-value": "_c", "data-type": "reset" }));
}
export const Button = Object.assign(ButtonRoot, {
    Link: ButtonLink,
    Mint: ButtonMint,
    Redirect: ButtonRedirect,
    Reset: ButtonReset,
});
function normalizeChildren(children) {
    return Array.isArray(children) ? children.join('') : children;
}
